Add first batch of alma8 CDTs

h-vetinari opened the pull request:
This tackles the first batch of CDTs from #66, more specifically those currently marked ✅ [here](https://github.com/conda-forge/cdt-builds/issues/66#issuecomment-1833417828). Additionally, I've added:
* The `-devel` packages from the dependencies we've added for [pam](https://github.com/conda-forge/cdt-builds/pull/55); IMO it makes sense to always have `libfoo` and `libfoo-devel` for a given CDT, otherwise `libfoo` is going to be non-functional (unless as a dependency for another CDT)
* `mesa-libgbm` because it was the only mesa-* package not to receive an ❌; I had [asked](https://github.com/conda-forge/cdt-builds/issues/66#issuecomment-2101919892) if that was intentional, Isuru had answered in the affirmative (but only in a core call, not as a comment in the issue)
* `kmod-devel` & `mesa-libgbm-devel` for the same reason; since we alread have `kmod` & `mesa-libgbm`

Doing this turned out to be a good exercise in the sense that the alma8 builds are separated across various subfolders that may depend on each other, which needed some updates to the logic.

Furthermore, I've switched away from `skipped_cdts` in favour of an allowlist-approach, as the impression I got from #66 was that people want to keep the list absolutely minimal, and it's easier to keep track of that using a positive list, rather than ensuring that we captured all exclusions correctly everywhere.

Finally, I repeatedly got permission errors (HTTP 403) when trying to `urlopen` https://vault.almalinux.org/8.9/BaseOS/{base_architecture}/os/repodata/repomd.xml, so I ended up switching to `requests` which doesn't have a problem there.

Given the large mechanical changes, it's probably easiest to review per commit.

--------------

Please see the repo readme for directions on how to make PRs on this repo.

Checklist:

- [x] if you have added a CDT, it appears in the `cdt_slugs.yaml` file and
  you have rerun the script `python gen_cdt_recipes.py`.
- [x] if you have changed the CDT generator code (`rpm.py`), ~you have bumped
  the build number in `conda_build_config.yaml`~ and have remade all of the
  recipes via running `python gen_cdt_recipes.py --force`
- [x] if you have added a custom CDT recipe, you have added the name of the CDT
  with `custom: true` in the `cdt_slugs.yaml` file.
- [x] all CDT recipes have build number set by `{{ cdt_build_number }}` for
  old-style/legacy CDTs or `{{ cdt_build_number|int + 1000 }}` for new-style CDTs
- [x] if you see a warning about a CDT not having a license, you have added the
  `license_file` key in the `cdt_slugs.yaml` file with the path to the appropriate
  license in `licenses/`

**NOTE: If you make any changes to `cd_slugs.yaml`, you need to reun the generator code
via `python gen_cdt_recipes.py`.**

isuruf commented on cdts/pam-devel-alma8-x86_64/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+package:
+  name: pam-devel-alma8-x86_64
```
Let's name this with `conda` instead of `alma8` as discussed in core meeting.

h-vetinari replied:
Done in https://github.com/conda-forge/cdt-builds/pull/70/commits/b42c06110e18016d294e37e0fe8ba641fa281fda ff.

beckermr reviewed (changes requested):
Let's put the commits of the new CDT recipes into another PR. I cannot review the new logic commit by commit.

h-vetinari commented:
> Let's put the commits of the new CDT recipes into another PR.

It's hard to validate the logic without actually rendering & building things... But if you prefer, I'll revert this at the end once it's green (or rebase it out) for ease of review.

> I cannot review the new logic commit by commit.

Aside from the one rerender commit, the diffs are tiny, and GH makes it easy to click through (prev/next)...?

beckermr commented:
I'm not currently figuring out how to do it for the commits but ok. I'll do my best.

beckermr reviewed (changes requested):
The yaml file had support for other things that had been important:

1. excluding certain deps (e.g., kernel-headers)
2. putting in custom logic in the build scripts for certain CDTs

When you added alma8 with the allow list, you didn't implement those features for alma8. How do you know we won't need them? Why not stick to the yaml so we can keep them?

beckermr commented:
One concrete way to do this would be to have a separate cdt slugs file for alma8 and cos7. Then each one is an allow list on its own and we can keep the extra hooks if we need them.

h-vetinari commented:
> The yaml file had support for other things that had been important:

None of the logic _except_ `skipped_cdts` is changed. Removing/replacing deps & custom logic is still possible

> Why not stick to the yaml so we can keep them?

I haven't changed away from the yaml.

beckermr commented:
So I have to list an alma8 CDT in both the yaml AND the allow list in the python file?

beckermr commented:
Maybe the better question is, can you show me how to add a skipped dependency for an alma8 CDT?

h-vetinari commented:
> One concrete way to do this would be to have a separate cdt slugs file for alma8 and cos7.

So, with the caveat that all of this is still a draft, I'd prefer not split the cdt slugs, because for all the ones we're interested (so far), the config is exactly the same as for cos7, so duplicating them would be kinda gratuitous.

FWIW, I had started with applying the `skipped_cdts` approach, and it just ends up with
```yaml
  skipped_cdts:
    - alma8-x86_64
    - alma8-aarch64
    - alma8-aarch64
```
for almost all slugs

h-vetinari commented:
> So I have to list an alma8 CDT in both the yaml AND the allow list in the python file?

If it's an entirely new CDT then yes. If it's a CDT that's already on cos7, then only the allowlist. Again, please consider this a suggestion (it seemed more natural given how much people want to constrain CDT creation), not as something set in stone.

h-vetinari commented on cdt_slugs.yaml:
```diff
@@ -6,6 +6,11 @@ alsa-lib-devel:
+audit-libs-devel:
+  custom: false
+  license_file: licenses/audit-libs-copying.lib
+  dep_remove:
+    - kernel-headers
```
> Maybe the better question is, can you show me how to add a skipped dependency for an alma8 CDT?

This works for both cos7/alma8

h-vetinari replied:
@beckermr, this case might actually justify some further functionality. My understanding is that `dep_replace` can only replace one CDT with _another_ CDT, but what we'd actually want here is to replace a CDT with a conda-package, because we do have `kernel-headers_<arch>` coming from https://github.com/conda-forge/linux-sysroot-feedstock. 

For now I've removed `kernel-headers` because even if `audit-libs-devel` depends on it, I didn't want to pull it into the alma8 allowlist. It's possible though that the package is non-functional without the headers.

h-vetinari replied:
> which always pulls in the kernel headers.

Yeah, the kernel-headers are special in that they're a dep of the sysroot which is there already. I guess we can defer this question (i.e. a CDT that depends on another distro-package for which we have a conda equivalent) if and until it ever becomes relevant.

beckermr commented:
OK thanks for answering my questions @h-vetinari!

Maybe we can refactor the cdt_slugs.yaml file to look like this:

```yaml
allowed_cdts:
  alma8:
    - foo
    - bar
    - blah
  cos7:
    - foo
    - bar
    - blah
    - glarg
cdt_build_defs:
  # any CDT not listed here is equivalent to the default
  #
  #   <cdt name>:
  #     custom: false
  #
  foo:
    license_file: licenses/alsa-lib-license
  bar:
    dep_remove:
      - foobarg
```

This way all of the configuration is still in one spot and we only have to list things that are special/different.

beckermr commented:
Yeah I am ok with the indentation change.

h-vetinari commented:
> I'm not currently figuring out how to do it for the commits but ok. I'll do my best.

FWIW, there's a commit tab on top (between "Conversation" and "Checks"); if you click on the first commit therein, you'll get a prev/next navigation menu in the upper RH corner. Aside from the rerender commits (where you can just click next without review), the others are very small and should be easy to digest.

> Yeah I am ok with the indentation change.

I might not finish this today, but it shouldn't be hard conceptually; for now I made the change that Isuru asked for, and it's getting late here.

beckermr commented:
No rush on this and thanks for the dev work @h-vetinari!

h-vetinari reviewed (commented):
Another infrastructure question

h-vetinari commented:
> This way all of the configuration is still in one spot and we only have to list things that are special/different.

From a quick check, only 30 CDTs out of 180 would actually fall under
```yaml
  # any CDT not listed here is equivalent to the default
  #
  #   <cdt name>:
  #     custom: false
```
It's a bit hard to argue that 83% are the special case IMO 😅

The vast majority of those only specifies the `license_file`, but unless we have a way to revamp the license handling too, I'm not sure this would be an improvement. It'd just be confusing look for a non-existent build definition of a CDT that you're interested in (e.g. often `libfoo-devel` needs a license override, but `libfoo` doesn't), only to realise you've fallen into the fallback case.

beckermr commented:
We can list them all then, that's fine by me. What doesn't work is having to add a CDT name in two places, once in a yaml file and once in the python code, in order to add a new CDT. That is confusing for sure.

h-vetinari commented:
@beckermr, I implemented your idea (minus the default fallback, as discussed), and a run of `python gen_cdt_recipes.py --force`  yields no further changes (as desired). I added some basic validation (tested to work locally) to fail early in case of missing `build_defs`, or warn in case of superfluous ones.

Feel free to play with this branch according to your usual workflows, happy to fix. :)

beckermr reviewed (approved):
LGTM! Thank you!

isuruf commented:
Why is there a java CDT? Is it a transitive dependency?

h-vetinari commented:
> Why is there a java CDT? Is it a transitive dependency?

I didn't touch those, they're pre-existing (also present in the list in #66 since last Nov.). I looked for current users and even checked now if they were used in any of the non-java packages we dropped in https://github.com/conda-forge/cdt-builds/commit/779da0bbf312f65a6d2a35e930abb599d98ac309, but can't find any users in the CDTs themselves.

Looking [around](https://github.com/search?q=org%3Aconda-forge+%2Fcdt%5C%28.*java%2F&type=code) conda-forge, it's [used](https://github.com/conda-forge/r-base-feedstock/blob/7259ce0272e70090e8d08f611995a1d0a71df958/recipe/meta.yaml#L75), but _only_ in `r-base`. Once you remove that, we can consider dropping them here.

isuruf commented:
Ah, it's just a whitespace change.

jakirkham commented on build_cdt_recipes.py:
```diff
@@ -290,8 +290,8 @@ def _main(
     """
-    Build all CDT recipes for a given DIST_ARCH_SLUG (e.g. cos6-x86_64,
-    cos7-aarch64, etc.)
+    Build all CDT recipes for a given DIST_ARCH_SLUG (e.g. conda-x86_64
+    for post-CentOS CDTs, or cos6-x86_64, cos7-aarch64, etc. historically)
```
Noticing that the new slugs don't have a version. Did we want to bake one in like we did with `cos6` & `cos7`? For example `conda-2_28-<arch>`

isuruf replied:
Yes. See your comment on https://github.com/conda-forge/conda-forge.github.io/blob/5e57e30a11e91dc127aab1543b2cffe8ed8b3290/community/minutes/2024-05-15.md?plain=1#L74-L77

jakirkham reviewed (commented):
Thanks all for getting this going! 🙏

Had a question below about naming/versioning
